api/pkg/apis/v1alpha1/providers/secret/k8sSecretProvider.go

```diff
@@ -0,0 +1,179 @@
+
+	sLog.Debug("  P (K8s Secret): initialize")
+
+	updateConfig, err := toK8sStateProviderConfig(config)
+	if err != nil {
+		sLog.Errorf("  P (K8s Secret): expected KubectlTargetProviderConfig: %+v", err)
+		return err
+	}
+	s.Config = updateConfig
+	var kConfig *rest.Config
+	if s.Config.InCluster {
+		kConfig, err = rest.InClusterConfig()
+	} else {
+		switch s.Config.ConfigType {
+		case "path":
+			if s.Config.ConfigData == "" {
+				if home := homedir.HomeDir(); home != "" {
+					s.Config.ConfigData = filepath.Join(home, ".kube", "config")
+				} else {
+					err = v1alpha2.NewCOAError(nil, "can't locate home direction to read default kubernetes config file, to run in cluster, set inCluster config setting to true", v1alpha2.BadConfig)
+					sLog.Errorf("  P (K8s State): %+v", err)
+					return err
+				}
+			}
+			kConfig, err = clientcmd.BuildConfigFromFlags("", s.Config.ConfigData)
+		case "bytes":
+			if s.Config.ConfigData != "" {
+				kConfig, err = clientcmd.RESTConfigFromKubeConfig([]byte(s.Config.ConfigData))
+				if err != nil {
+					sLog.Errorf("  P (K8s State): %+v", err)
+					return err
+				}
+			} else {
+				err = v1alpha2.NewCOAError(nil, "config data is not supplied", v1alpha2.BadConfig)
+				sLog.Errorf("  P (K8s State): %+v", err)
+				return err
+			}
+		default:
+			err = v1alpha2.NewCOAError(nil, "unrecognized config type, accepted values are: path and bytes", v1alpha2.BadConfig)
+			sLog.Errorf("  P (K8s State): %+v", err)
+			return err
+		}
+	}
+	if err != nil {
+		sLog.Errorf("  P (K8s State): %+v", err)
+		return err
+	}
+	s.Clientset, err = kubernetes.NewForConfig(kConfig)
+	if err != nil {
+		sLog.Errorf("  P (K8s State): %+v", err)
+		return err
+	}
+
+	return nil
+}
+
+func toK8sStateProviderConfig(config providers.IProviderConfig) (K8sSecretProviderConfig, error) {
+	ret := K8sSecretProviderConfig{}
+	data, err := json.Marshal(config)
+	if err != nil {
+		return ret, err
+	}
+	err = json.Unmarshal(data, &ret)
+	if ret.ConfigType == "" {
+		ret.ConfigType = "path"
+	}
+	return ret, err
+}
+
+func (s *K8sSecretProvider) Read(name string, field string, localContext interface{}) (string, error) {
+	// Get the secret
+	namespace := s.getNamespaceFromContext(localContext)
+	secret, err := s.Clientset.CoreV1().Secrets(namespace).Get(context.Background(), name, metav1.GetOptions{})
+	if err != nil {
+		sLog.Errorf("Error getting secret %s in namespace %s. Error: %s", name, namespace, err.Error())
+		return "", err
+	}
+	// Get the field from the secret data
+	value, ok := secret.Data[field]
+	if !ok {
+		sLog.Errorf("Field %s not found in secret %s", field, name)
+		return "", fmt.Errorf("field %s not found in secret %s", field, name)
+	}
+
+	return string(value), nil
+}
+
+func (s *K8sSecretProvider) getNamespaceFromContext(localContext interface{}) string {
+	if localContext != nil {
+		if ltx, ok := localContext.(coa_utils.EvaluationContext); ok {
+			return ltx.Namespace
+		}
+	}
+	return " "
```

Comment:
could you please extract a common function as it's used by both config and secret?